lib/ibandit/iban.rb

```diff
@@ -120,17 +87,57 @@ def valid_check_digits?
       end
     end
 
-    def valid_length?
+    def valid_bank_code_length?
       return unless valid_country_code?
 
-      if iban.size == structure[:total_length]
-        true
-      else
-        @errors[:length] = "Length doesn't match SWIFT specification " \
-                           "(expected #{structure[:total_length]} " \
-                           "characters, received #{iban.size})"
-        false
+      if bank_code.nil?
+        @errors[:bank_code] = 'is required'
+        return false
       end
+
+      return true if bank_code.length == structure[:bank_code_length]
+
+      @errors[:bank_code] = 'is the wrong length: must be ' \
+                            "#{structure[:bank_code_length]}, not " \
+                            "#{bank_code.length}"
+      false
+    end
+
+    def valid_branch_code_length?
+      return unless valid_country_code?
+
+      if branch_code.length == 0 && structure[:branch_code_length] > 0
```

Comment:
Actually, I need to remove all the `nil` stuff - `branch_code` etc are defined with `... || ''` so will have empty string instead of `nil` to avoid all the extra cases.